app/src/main/java/com/dessalines/thumbkey/utils/Utils.kt

```diff
@@ -941,3 +942,96 @@ fun getLocalCurrency(): String? {
         ).currency?.symbol
     }
 }
+
+fun makeVariant(
+    source: KeyboardC,
+    variant: KeyboardC,
+): KeyboardC {
+    // Returns a variant of a layer created by replacing source keys and swipes
+    // with any non-dummy equivalents in the variant.
+
+    // Assuming both source and variant have the same structure (same number of rows and columns)
+    val newRows =
+        source.arr.indices.map { rowIndex ->
+            source.arr[rowIndex].indices.map { colIndex ->
+                val sourceKeyItem = source.arr[rowIndex][colIndex]
+                val variantKeyItem = variant.arr[rowIndex][colIndex]
+                mergeKeyItems(sourceKeyItem, variantKeyItem)
+            }
+        }
+    return source.copy(arr = newRows)
+}
+
+fun mergeKeyItems(
+    source: KeyItemC,
+    variant: KeyItemC,
+): KeyItemC {
+    // Returns merge of keyItems.
+    // If source has swipes of same direction as variant, they'll be replaced.
+    // Same with center, unless it's a dummy key (since a centre key is
+    // mandatory).
+
+    // Determine the new center. If the variant's center is a 'dummy', keep the source's center.
+    val newCenter = if (variant.center == DUMMY_KEY) source.center else variant.center
+
+    // Merge swipes, with variant's swipes overriding source's swipes for the same SwipeDirection
+    val newSwipes =
+        source.swipes.orEmpty().toMutableMap().apply {
```

Comment:
`apply` might be the issue here, I think it alters source data.

Reply:
Ah, I thought I was creating a mutable deep copy. I'll investigate, thanks

Reply:
no dice so far with this route of investigation. converting to a map creates a copy, so the original isn't being modified at all

Reply:
Not looking like this was the issue. Probably need to figure out how to apply the debugger. Gonna be a wee while, I might not get time to work on this for a week or two